libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -3274,6 +3282,67 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::bytesToFixedBytesConversionFunction(ArrayType const& _from, FixedBytesType const& _to)
+{
+	solAssert(_from.isByteArray(), "");
+	solAssert(_from.isDynamicallySized(), "");
+	string functionName = "convert_bytes_to_bytesNN_from_" + _from.identifier() + "_to_" + _to.identifier();
+	return m_functionCollector.createFunction(functionName, [&](auto& _args, auto& _returnParams) {
+		_args = { "array" };
+		bool fromCalldata = _from.dataStoredIn(DataLocation::CallData);
+		if (fromCalldata)
+			_args.emplace_back("len");
+		_returnParams = {"value"};
+		Whiskers templ(R"(
+			let length := <arrayLen>(array<?fromCalldata>, len</fromCalldata>)
+			if gt(length, <fixedBytesLen>) { <panic>() }
+			let dataArea := array
+			<?fromMemory>
+				dataArea := <dataArea>(array)
+			</fromMemory>
+			<?fromStorage>
+				if eq(length, 32) { dataArea := <dataArea>(array) }
+			</fromStorage>
+
+			<?fromCalldata>
+				value := <cleanup>(calldataload(dataArea))
+			<!fromCalldata>
+				value := <extractValue>(dataArea)
+			</fromCalldata>
+
+			if lt(length, <fixedBytesLen>) {
+				value := and(
+					value,
+					<shl>(
+						mul(8, sub(<fixedBytesLen>, length)),
+						<mask>
+					)
+				)
+			}
+		)");
+		templ("fromCalldata", fromCalldata);
+		templ("arrayLen", arrayLengthFunction(_from));
+		templ("fixedBytesLen", to_string(_to.numBytes()));
+		templ("panic", panicFunction(PanicCode::FixedBytesConversionError));
+		templ("fromMemory", _from.dataStoredIn(DataLocation::Memory));
+		templ("fromStorage", _from.dataStoredIn(DataLocation::Storage));
+		templ("dataArea", arrayDataAreaFunction(_from));
+		if (fromCalldata)
+			templ("cleanup", cleanupFunction(_to));
+		else
+			templ(
+				"extractValue",
+				_from.dataStoredIn(DataLocation::Storage) ?
+				readFromStorage(_to, 32 - _to.numBytes(), true) :
```

Comment:
```suggestion
				readFromStorage(_to, 32 - _to.numBytes(), false) :
```
(creates a shorter function name)

Reply:
Also why is offset not zero?

Reply:
Zero index byte in storage bytes array is at offset 31, and length at 0 offset if I am correct

Reply:
ah right!